aws/resource_aws_kms_grant.go

```diff
@@ -0,0 +1,502 @@
+
+	err := resource.Retry(3*time.Minute, func() *resource.RetryError {
+		var err error
+
+		out, err = conn.CreateGrant(&input)
+
+		if err != nil {
+			if awsErr, ok := err.(awserr.Error); ok {
+				// Error Codes: https://docs.aws.amazon.com/sdk-for-go/api/service/kms/#KMS.CreateGrant
+				// Under some circumstances a newly created IAM Role doesn't show up and causes
+				// an InvalidArnException to be thrown.
+				if awsErr.Code() == "DependencyTimeoutException" ||
+					awsErr.Code() == "InternalException" ||
+					awsErr.Code() == "InvalidArnException" {
+					return resource.RetryableError(
+						fmt.Errorf("[WARN] Error adding new KMS Grant for key: %s, retrying %s",
+							*input.KeyId, err))
+				}
+			}
+			log.Printf("[ERROR] An error occured creating new AWS KMS Grant: %s", err)
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Created new KMS Grant: %s", *out.GrantId)
+	d.SetId(*out.GrantId)
+	d.Set("grant_id", out.GrantId)
+	d.Set("grant_token", out.GrantToken)
+
+	return resourceAwsKmsGrantRead(d, meta)
+}
+
+func resourceAwsKmsGrantRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kmsconn
+
+	grantId := d.Id()
+	keyId := d.Get("key_id").(string)
+
+	log.Printf("[DEBUG] Looking for grant id: %s", grantId)
+	grant, err := findKmsGrantByIdWithRetry(conn, keyId, grantId)
+
+	if err != nil {
+		return err
+	}
+
+	if grant == nil {
+		log.Printf("[WARN] %s KMS grant id not found for key id %s, removing from state file", grantId, keyId)
+		d.SetId("")
+	}
+
+	if grant != nil {
+		// The grant sometimes contains principals that identified by their unique id: "AROAJYCVIVUZIMTXXXXX"
+		// instead of "arn:aws:...", in this case don't update the state file
+		if strings.HasPrefix(*grant.GranteePrincipal, "arn:aws") {
+			d.Set("grantee_principal", grant.GranteePrincipal)
+		} else {
+			log.Printf(
+				"[WARN] Unable to update grantee principal state %s for grant id %s for key id %s.",
+				*grant.GranteePrincipal, grantId, keyId)
+		}
+
+		if grant.RetiringPrincipal != nil {
+			if strings.HasPrefix(*grant.RetiringPrincipal, "arn:aws") {
+				d.Set("retiring_principal", grant.RetiringPrincipal)
+			} else {
+				log.Printf(
+					"[WARN] Unable to update retiring principal state %s for grant id %s for key id %s",
+					*grant.RetiringPrincipal, grantId, keyId)
+			}
+		}
+
+		if err := d.Set("operations", aws.StringValueSlice(grant.Operations)); err != nil {
+			log.Printf("[DEBUG] Error setting operations for grant %s with error %s", grantId, err)
+		}
+		if *grant.Name != "" {
+			d.Set("name", grant.Name)
+		}
+		if grant.Constraints != nil {
+			if err := d.Set("constraints", flattenKmsGrantConstraints(grant.Constraints)); err != nil {
+				log.Printf("[DEBUG] Error setting constraints for grant %s with error %s", grantId, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// Retiring grants requires special permissions (i.e. the
```

Comment:
Seems like this behavior should be noted in the documentation. Better yet, it would probably be best to provide an attribute flag to control the behavior (e.g. `retire_on_delete`) that defaults to `true` so we can always return an error if an expected deletion (retire or just revoke) is not possible.

Reply:
Done. However, I inverted the default to false. Functionally retiring and revoking are the same, they don't show up in the ListGrants call, however retiring requires special permissions. It seemed to me that defaulting to that would be noisy and cause confusion.